docs/hacking.md

```diff
@@ -176,3 +176,150 @@ files that serve different purposes in the compiler:
+
+- Step 2. Lookup binders in the current context
+
+  Assuming `immutable` gets a correct argument, our next step is to use the
+  `Sym` XObj it received to find out if the symbol is bound to a variable or not.
+
+  `Lookup.hs` defines functions for looking up bindings in the various
+  environments contained in a context. We'll call lookup functions to check
+  whether or not the symbol argument we get is bound to a `def` form (in which
+  case it's a variable). If the symbol isn't bound to a `def` we'll error.
+
+
+  So, we'll get the binding for our argument (a `Binder`), match against the
+  binding's `XObj` and continue working only if it's a `def`.
+
+  ```
+  primitiveImmutable :: Primitive -- our new primitive
+  primitiveImmutable xobj ctx [XObj (Sym path@(SymPath) _)] =
+    let global = contextGlobalEnv ctx
+        binding = lookupInEnv path global
+    in  case binding of
+          Just (_, Binder meta (XObj )) -> -- TODO: This is a def! Great. Do more work here.
+          _ -> -- anything that isn't a def; error
+            return $ evalError ctx ("`immutable` expects a variable as an argument") (info xobj)
+  primitiveImmutable _ _ xobjs = -- any other number or types of xobj arguments are incorrect! Let's error.
+    return $ evalError ctx ("`immutable` expected a single symbol argument, but got" ++ show xobjs) (info xobj)
+  ```
+
+- Step 3. Perform logic; update the `Context`
+
+  Finally, now that we're certain we've got a def, we'll just perform our
+  special logic then update the context with our modified binder.
+
+  To keep things simple, all we'll do in this primitive is update the binder's
+  `MetaData` with a new key called `immutable` set to `true`. We can later use
+  the value of this meta field to prevent calls to `set!`.
+
+  ```
+  primitiveImmutable :: Primitive -- our new primitive
+  primitiveImmutable xobj ctx [XObj (Sym path@(SymPath) _)] =
+    let global = contextGlobalEnv ctx
+        binding = lookupInEnv path global
+    in  case binding of
+          Just (_, Binder meta def@(XObj Def _ _)) ->
+            let oldMeta = getMeta meta
+                newMeta = meta {getMeta = Map.insert "immutable" trueXObj oldMeta}  -- update the binder metadata
+            in  return $ ctx {contextGlobalEnv = Env (envInsertAt global path (Binder newMeta def))} -- update the context with the binder and it's new meta and return
+          _ -> -- anything that isn't a def; error
+            return $ evalError ctx ("`immutable` expects a variable as an argument") (info xobj)
+  primitiveImmutable _ _ xobjs = -- any other number or types of xobj arguments are incorrect! Let's error.
+    return $ evalError ctx ("`immutable` expected a single symbol argument, but got" ++ show xobjs) (info xobj)
+  ```
+
+And that wraps up the core logic of our primitive. To make it available, we
+just need to register it in `StartingEnv.hs`.
+
+
+#### 1. Add your primitive to the starting environment
```

Comment:
Why is this heading numbered?There do not seem to be any other numbers anywhere?

Reply:
There is another numbered heading above, but I think it makes sense just to drop the numbers entirely (I numbered this one wrong anyways).